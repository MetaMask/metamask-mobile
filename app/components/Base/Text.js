import React from 'react';
import PropTypes from 'prop-types';
import { Text as RNText, StyleSheet } from 'react-native';
import { fontStyles, colors } from '../../styles/common';

const style = StyleSheet.create({
	text: {
		...fontStyles.normal,
		color: colors.grey600,
		marginVertical: 2,
		fontSize: 14
	},
	centered: {
		textAlign: 'center'
	},
	right: {
		textAlign: 'right'
	},
	bold: fontStyles.bold,
	green: {
		color: colors.green400
	},
	primary: {
		color: colors.fontPrimary
	},
	small: {
		fontSize: 12
	},
	upper: {
		textTransform: 'uppercase'
	},
	disclaimer: {
		fontStyle: 'italic',
		letterSpacing: 0.15
	},
	modal: {
		color: colors.fontPrimary,
		fontSize: 16,
		lineHeight: 30
	},
	link: {
		color: colors.blue
	},
	strikethrough: {
		textDecorationLine: 'line-through'
	}
});

const Text = ({
	reset,
	centered,
	right,
	bold,
	green,
	primary,
	small,
	upper,
	modal,
	disclaimer,
	link,
	strikethrough,
	style: externalStyle,
	...props
}) => (
	<RNText
		style={[
			!reset && style.text,
			centered && style.centered,
			right && style.right,
			bold && style.bold,
			green && style.green,
			primary && style.primary,
			disclaimer && [style.small, style.disclaimer],
			small && style.small,
			upper && style.upper,
			modal && style.modal,
			link && style.link,
			strikethrough && style.strikethrough,
			externalStyle
		]}
		{...props}
	/>
);

Text.defaultProps = {
	reset: false,
	centered: false,
	right: false,
	bold: false,
	green: false,
	primary: false,
	disclaimer: false,
	modal: false,
	small: false,
	upper: false,
	link: false,
	strikethrough: false,
	style: undefined
};

Text.propTypes = {
	/**
	 * Removes teh default style
	 */
	reset: PropTypes.bool,
	/**
	 * Align text to center
	 */
	centered: PropTypes.bool,
	/**
	 * Align text to right
	 */
	right: PropTypes.bool,
	/**
	 * Makes text bold
	 */
	bold: PropTypes.bool,
	/**
	 * Makes text green
	 */
	green: PropTypes.bool,
	/**
	 * Makes text fontPrimary color
	 */
	primary: PropTypes.bool,
	/**
	 * Makes text italic and tight
	 * used in disclaimers
	 */
	disclaimer: PropTypes.bool,
	/**
	 * Makes text black and bigger
	 * Used in modals
	 */
	modal: PropTypes.bool,
	/**
	 * Makes text small
	 */
	small: PropTypes.bool,
	/**
	 * Makes text uppercase
	 */
	upper: PropTypes.bool,
	/**
	 * Applies a link style
	 */
	link: PropTypes.bool,
	/**
	 * Applies a strikethrough decoration
	 */
	strikethrough: PropTypes.bool,
	/**
	 * Any other external style defined in props will be applied
	 */
	style: PropTypes.oneOfType([PropTypes.object, PropTypes.array])
};

export default Text;
