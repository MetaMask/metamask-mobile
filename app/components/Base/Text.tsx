import React from 'react';
import { Text as RNText, StyleSheet } from 'react-native';
import { fontStyles } from '../../styles/common';
import { useTheme } from '../../util/theme';
import { Colors } from '../../util/theme/models';

interface TextProps extends React.ComponentPropsWithoutRef<typeof RNText> {
  reset?: boolean;
  centered?: boolean;
  right?: boolean;
  bold?: boolean;
  green?: boolean;
  black?: boolean;
  blue?: boolean;
  red?: boolean;
  grey?: boolean;
  orange?: boolean;
  infoModal?: boolean;
  noMargin?: boolean;
  primary?: boolean;
  muted?: boolean;
  disclaimer?: boolean;
  modal?: boolean;
  small?: boolean;
  big?: boolean;
  bigger?: boolean;
  upper?: boolean;
  link?: boolean;
  strikethrough?: boolean;
  underline?: boolean;
}

const createStyles = (colors: Colors) =>
  StyleSheet.create({
    text: {
      ...fontStyles.normal,
      color: colors.text.default,
      marginVertical: 2,
      fontSize: 14,
    },
    centered: {
      textAlign: 'center',
    },
    right: {
      textAlign: 'right',
    },
    red: {
      color: colors.error.default,
    },
    orange: {
      color: colors.secondary.default,
    },
    black: {
      color: colors.text.default,
    },
    bold: fontStyles.bold,
    blue: {
      color: colors.primary.default,
    },
    green: {
      color: colors.success.default,
    },
    grey: {
      color: colors.text.alternative,
    },
    primary: {
      color: colors.text.default,
    },
    muted: {
      color: colors.text.muted,
    },
    small: {
      fontSize: 12,
    },
    big: {
      fontSize: 16,
    },
    bigger: {
      fontSize: 18,
    },
    upper: {
      textTransform: 'uppercase',
    },
    disclaimer: {
      fontStyle: 'italic',
      letterSpacing: 0.15,
    },
    modal: {
      color: colors.text.default,
      fontSize: 16,
      lineHeight: 22.4, // 1.4 * fontSize
    },
    infoModal: {
      lineHeight: 20,
      marginVertical: 6,
    },
    link: {
      color: colors.primary.default,
    },
    strikethrough: {
      textDecorationLine: 'line-through',
    },
    underline: {
      textDecorationLine: 'underline',
    },
    noMargin: {
      marginVertical: 0,
    },
  });

const Text: React.FC<TextProps> = ({
  reset,
  centered,
  right,
  bold,
  green,
  black,
  blue,
  grey,
  red,
  orange,
  primary,
  muted,
  small,
  big,
  bigger,
  upper,
  modal,
  infoModal,
  disclaimer,
  link,
  strikethrough,
  underline,
  noMargin,
  style: externalStyle,
  ...props
}: TextProps) => {
  const { colors } = useTheme();
  const style = createStyles(colors);

  return (
    <RNText
      style={[
        !reset && style.text,
        centered && style.centered,
        right && style.right,
        bold && style.bold,
        green && style.green,
        black && style.black,
        blue && style.blue,
        grey && style.grey,
        red && style.red,
        orange && style.orange,
        black && style.black,
        primary && style.primary,
        muted && style.muted,
        disclaimer && [style.small, style.disclaimer],
        small && style.small,
        big && style.big,
        bigger && style.bigger,
        upper && style.upper,
        modal && style.modal,
        infoModal && style.infoModal,
        link && style.link,
        strikethrough && style.strikethrough,
        underline && style.underline,
        noMargin && style.noMargin,
        externalStyle,
      ]}
      {...props}
    />
  );
};

export default Text;
