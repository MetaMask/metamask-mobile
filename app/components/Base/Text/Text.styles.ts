import { StyleSheet } from 'react-native';
import { scale } from 'react-native-size-matters';
import { fontStyles } from '../../../styles/common';
import { Colors } from '../../../util/theme/models';

const createStyles = (colors: Colors) =>
  StyleSheet.create({
    text: {
      ...fontStyles.normal,
      color: colors.text.default,
      marginVertical: 2,
      fontSize: scale(14),
    },
    centered: {
      textAlign: 'center',
    },
    right: {
      textAlign: 'right',
    },
    red: {
      color: colors.error.default,
    },
    orange: {
      color: colors.secondary.default,
    },
    black: {
      color: colors.text.default,
    },
    bold: fontStyles.bold,
    blue: {
      color: colors.primary.default,
    },
    green: {
      color: colors.success.default,
    },
    grey: {
      color: colors.text.alternative,
    },
    primary: {
      color: colors.text.default,
    },
    muted: {
      color: colors.text.muted,
    },
    small: {
      fontSize: 12,
    },
    big: {
      fontSize: 16,
    },
    bigger: {
      fontSize: 18,
    },
    upper: {
      textTransform: 'uppercase',
    },
    disclaimer: {
      fontStyle: 'italic',
      letterSpacing: 0.15,
    },
    modal: {
      color: colors.text.default,
      fontSize: 16,
      lineHeight: 22.4, // 1.4 * fontSize
    },
    infoModal: {
      lineHeight: 20,
      marginVertical: 6,
    },
    link: {
      color: colors.primary.default,
    },
    strikethrough: {
      textDecorationLine: 'line-through',
    },
    underline: {
      textDecorationLine: 'underline',
    },
    noMargin: {
      marginVertical: 0,
    },
  });

export default createStyles;
