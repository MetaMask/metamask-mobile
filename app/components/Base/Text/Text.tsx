import React from 'react';
import { Text as RNText } from 'react-native';
import { useTheme } from '../../../util/theme';
import styles from './Text.styles';

interface TextProps extends React.ComponentPropsWithoutRef<typeof RNText> {
  reset?: boolean;
  centered?: boolean;
  right?: boolean;
  bold?: boolean;
  green?: boolean;
  black?: boolean;
  blue?: boolean;
  red?: boolean;
  grey?: boolean;
  orange?: boolean;
  infoModal?: boolean;
  noMargin?: boolean;
  primary?: boolean;
  muted?: boolean;
  disclaimer?: boolean;
  modal?: boolean;
  small?: boolean;
  big?: boolean;
  bigger?: boolean;
  upper?: boolean;
  link?: boolean;
  strikethrough?: boolean;
  underline?: boolean;
}

const Text: React.FC<TextProps> = ({
  reset,
  centered,
  right,
  bold,
  green,
  black,
  blue,
  grey,
  red,
  orange,
  primary,
  muted,
  small,
  big,
  bigger,
  upper,
  modal,
  infoModal,
  disclaimer,
  link,
  strikethrough,
  underline,
  noMargin,
  style: externalStyle,
  ...props
}: TextProps) => {
  const { colors } = useTheme();
  const style = styles(colors);

  return (
    <RNText
      style={[
        !reset && style.text,
        centered && style.centered,
        right && style.right,
        bold && style.bold,
        green && style.green,
        black && style.black,
        blue && style.blue,
        grey && style.grey,
        red && style.red,
        orange && style.orange,
        black && style.black,
        primary && style.primary,
        muted && style.muted,
        disclaimer && [style.small, style.disclaimer],
        small && style.small,
        big && style.big,
        bigger && style.bigger,
        upper && style.upper,
        modal && style.modal,
        infoModal && style.infoModal,
        link && style.link,
        strikethrough && style.strikethrough,
        underline && style.underline,
        noMargin && style.noMargin,
        externalStyle,
      ]}
      {...props}
    />
  );
};

export default Text;
