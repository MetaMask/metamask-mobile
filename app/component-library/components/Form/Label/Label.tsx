/* eslint-disable react/prop-types */

// Third party dependencies.
import React from 'react';

// External dependencies.
import Text from '../../Texts/Text';

// Internal dependencies.
import { LabelProps } from './Label.types';
import { DEFAULT_LABEL_TEXT_VARIANT, LABEL_TEST_ID } from './Label.constants';

const Label: React.FC<LabelProps> = ({ ...props }) => (
  <Text
    variant={DEFAULT_LABEL_TEXT_VARIANT}
    testID={LABEL_TEST_ID}
    {...props}
  />
);

export default Label;
