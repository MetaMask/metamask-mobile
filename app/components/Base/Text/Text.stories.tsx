import React from 'react';

import { storiesOf } from '@storybook/react-native';
import { action } from '@storybook/addon-actions';
import { text, boolean } from '@storybook/addon-knobs';

import Text from './Text';

storiesOf('Components / Base / Text', module)
  .addDecorator((getStory) => getStory())
  .add('Default', () => (
    <Text
      onPress={action('onPress')}
      reset={boolean('reset', false)}
      centered={boolean('centered', false)}
      right={boolean('right', false)}
      bold={boolean('bold', false)}
      green={boolean('green', false)}
      black={boolean('black', false)}
      blue={boolean('blue', false)}
      grey={boolean('grey', false)}
      red={boolean('red', false)}
      orange={boolean('orange', false)}
      primary={boolean('primary', false)}
      disclaimer={boolean('disclaimer', false)}
      small={boolean('small', false)}
      big={boolean('big', false)}
      upper={boolean('upper', false)}
      modal={boolean('modal', false)}
      infoModal={boolean('infoModal', false)}
      link={boolean('link', false)}
      strikethrough={boolean('strikethrough', false)}
      underline={boolean('underline', false)}
      noMargin={boolean('noMargin', false)}
    >
      {text('children', 'This is a Text component')}
    </Text>
  ));
