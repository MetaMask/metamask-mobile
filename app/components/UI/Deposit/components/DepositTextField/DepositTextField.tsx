import React, { forwardRef } from 'react';
import { View, StyleSheet } from 'react-native';
import { useStyles } from '../../../../hooks/useStyles';
import Label from '../../../../../component-library/components/Form/Label';
import Text, {
  TextVariant,
} from '../../../../../component-library/components/Texts/Text';
import TextField, {
  TextFieldSize,
} from '../../../../../component-library/components/Form/TextField';

interface DepositTextFieldProps {
  label: string;
  placeholder: string;
  value: string;
  onChangeText: (text: string) => void;
  error?: string;
  returnKeyType?: 'done' | 'next' | 'go' | 'search' | 'send';
  autoCapitalize?: 'none' | 'sentences' | 'words' | 'characters';
  keyboardType?:
    | 'default'
    | 'email-address'
    | 'numeric'
    | 'phone-pad'
    | 'number-pad';
  secureTextEntry?: boolean;
  testID?: string;
  style?: object;
  containerStyle?: object;
}

const styleSheet = () =>
  StyleSheet.create({
    subtitle: {
      marginBottom: 20,
    },
    label: {
      marginBottom: 6,
    },
    field: {
      flexDirection: 'column',
      marginBottom: 16,
    },
    error: {
      color: 'red',
      fontSize: 12,
      marginTop: 4,
    },
    errorPlaceholder: {
      height: 16, // Ensures consistent spacing even when no error
    },
  });

const DepositTextField = forwardRef<any, DepositTextFieldProps>(
  (
    {
      label,
      placeholder,
      value,
      onChangeText,
      error,
      returnKeyType = 'done',
      autoCapitalize = 'none',
      keyboardType = 'default',
      secureTextEntry = false,
      testID,
      style = {},
      containerStyle = {},
    },
    ref,
  ) => {
    const { styles, theme } = useStyles(styleSheet, {});

    return (
      <View style={[styles.field, containerStyle]}>
        <Label variant={TextVariant.HeadingSMRegular} style={styles.label}>
          {label}
        </Label>
        <TextField
          size={TextFieldSize.Lg}
          placeholder={placeholder}
          placeholderTextColor={theme.colors.text.muted}
          returnKeyType={returnKeyType}
          autoCapitalize={autoCapitalize}
          ref={ref}
          onChangeText={onChangeText}
          value={value}
          keyboardType={keyboardType}
          secureTextEntry={secureTextEntry}
          keyboardAppearance={theme.themeAppearance}
          testID={testID}
          style={style}
        />
        {error ? (
          <Text style={styles.error}>{error}</Text>
        ) : (
          <View style={styles.errorPlaceholder} />
        )}
      </View>
    );
  },
);

export default DepositTextField;
