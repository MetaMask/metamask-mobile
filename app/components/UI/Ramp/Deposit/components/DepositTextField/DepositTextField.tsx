import React, { forwardRef } from 'react';
import {
  View,
  StyleSheet,
  TextInput,
  ViewStyle,
  StyleProp,
} from 'react-native';
import { useStyles } from '../../../../../hooks/useStyles';
import Label from '../../../../../../component-library/components/Form/Label';
import Text, {
  TextVariant,
} from '../../../../../../component-library/components/Texts/Text';
import TextField, {
  TextFieldSize,
} from '../../../../../../component-library/components/Form/TextField';
import { TextFieldProps } from '../../../../../../component-library/components/Form/TextField/TextField.types';
import { Theme } from '../../../../../../util/theme/models';

interface DepositTextFieldProps extends Omit<TextFieldProps, 'size'> {
  label: string;
  error?: string;
  containerStyle?: StyleProp<ViewStyle>;
}

const styleSheet = (params: { theme: Theme }) => {
  const { theme } = params;

  return StyleSheet.create({
    label: {
      marginBottom: 6,
    },
    field: {
      flexDirection: 'column',
      marginBottom: 16,
    },
    error: {
      color: theme.colors.error.default,
      fontSize: 12,
      marginTop: 4,
    },
  });
};

const DepositTextField = forwardRef<TextInput, DepositTextFieldProps>(
  ({ label, error, containerStyle, style, ...textFieldProps }, ref) => {
    const { styles, theme } = useStyles(styleSheet, {});

    return (
      <View style={[styles.field, containerStyle]}>
        <Label variant={TextVariant.BodyMD} style={styles.label}>
          {label}
        </Label>
        <TextField
          size={TextFieldSize.Lg}
          placeholderTextColor={theme.colors.text.muted}
          keyboardAppearance={theme.themeAppearance}
          style={style}
          ref={ref}
          {...textFieldProps}
        />
        {error ? <Text style={styles.error}>{error}</Text> : null}
      </View>
    );
  },
);

export default DepositTextField;
