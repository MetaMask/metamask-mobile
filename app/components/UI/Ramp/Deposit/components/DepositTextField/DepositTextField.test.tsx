import React from 'react';
import { View } from 'react-native';
import { shallow } from 'enzyme';
import DepositTextField from './DepositTextField';
import Label from '../../../../../../component-library/components/Form/Label';
import Text, {
  TextVariant,
} from '../../../../../../component-library/components/Texts/Text';
import TextField, {
  TextFieldSize,
} from '../../../../../../component-library/components/Form/TextField';

const DEPOSIT_FIELD_TEST_ID = 'deposit-field-test-id';

const mockTheme = {
  colors: {
    text: { muted: '#888888' },
    error: { default: '#FF0000' },
  },
  themeAppearance: 'light',
};

const defaultProps = {
  label: 'Test Label',
  testID: DEPOSIT_FIELD_TEST_ID,
};

jest.mock('../../../../../hooks/useStyles', () => ({
  useStyles: () => ({
    styles: {
      field: {},
      label: {},
      error: {},
    },
    theme: mockTheme,
  }),
}));

describe('DepositTextField', () => {
  it('should render default settings correctly', () => {
    const wrapper = shallow(<DepositTextField {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render DepositTextField with correct label', () => {
    const wrapper = shallow(<DepositTextField {...defaultProps} />);
    const labelComponent = wrapper.find(Label);
    expect(labelComponent.exists()).toBe(true);
    expect(labelComponent.prop('variant')).toBe(TextVariant.HeadingSMRegular);
    expect(labelComponent.prop('children')).toBe('Test Label');
  });

  it('should render TextField component with correct props', () => {
    const wrapper = shallow(<DepositTextField {...defaultProps} />);
    const textFieldComponent = wrapper.find(TextField);
    expect(textFieldComponent.exists()).toBe(true);
    expect(textFieldComponent.prop('size')).toBe(TextFieldSize.Lg);
    expect(textFieldComponent.prop('placeholderTextColor')).toBe(
      mockTheme.colors.text.muted,
    );
    expect(textFieldComponent.prop('keyboardAppearance')).toBe(
      mockTheme.themeAppearance,
    );
  });

  it('should render error text when error prop is provided', () => {
    const errorMessage = 'This is an error message';
    const wrapper = shallow(
      <DepositTextField {...defaultProps} error={errorMessage} />,
    );
    const errorComponent = wrapper.find(Text);
    expect(errorComponent.exists()).toBe(true);
    expect(errorComponent.prop('children')).toBe(errorMessage);
  });

  it('should not render error text when error prop is not provided', () => {
    const wrapper = shallow(<DepositTextField {...defaultProps} />);
    const errorComponent = wrapper.find(Text);
    expect(errorComponent.exists()).toBe(false);
  });

  it('should apply custom container style when provided', () => {
    const customStyle = { marginTop: 20 };
    const wrapper = shallow(
      <DepositTextField {...defaultProps} containerStyle={customStyle} />,
    );
    const container = wrapper.find(View).first();
    const containerStyle = container.prop('style');
    expect(containerStyle).toEqual(
      expect.arrayContaining([expect.objectContaining(customStyle)]),
    );
  });

  it('should pass additional props to TextField', () => {
    const placeholder = 'Enter your text here';
    const maxLength = 50;
    const wrapper = shallow(
      <DepositTextField
        {...defaultProps}
        placeholder={placeholder}
        maxLength={maxLength}
      />,
    );
    const textFieldComponent = wrapper.find(TextField);
    expect(textFieldComponent.prop('placeholder')).toBe(placeholder);
    expect(textFieldComponent.prop('maxLength')).toBe(maxLength);
  });
});
