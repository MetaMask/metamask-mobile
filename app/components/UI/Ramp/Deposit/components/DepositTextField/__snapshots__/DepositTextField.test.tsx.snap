// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DepositTextField should render default settings correctly 1`] = `
<View
  style={
    [
      {},
      undefined,
    ]
  }
>
  <Label
    style={{}}
    variant="sHeadingSMRegular"
  >
    Test Label
  </Label>
  <ForwardRef
    keyboardAppearance="light"
    placeholderTextColor="#888888"
    size="48"
    testID="deposit-field-test-id"
  />
</View>
`;
